### PR TITLE
[refactor] Refactor Identifier::id_counter to global and local counter

### PR DESCRIPTION
**Pull request by PGZXB**
Related issue = #4577

### CODE REVIEW

File: taichi/ir/ir.h [hunk 1]
```diff
@@ -88,6 +88,9 @@ class Identifier {
     id = id_counter++;
```
Comment (k-ye) [inline taichi/ir/ir.h:88]: Should we remove `id_counter` later?

Comment (PGZXB) [reply taichi/ir/ir.h:88]: In the pr, the `is_counter` counts global identifier. I think that we can move `id_counter` into `Program` as member of it. WDYT?

Comment (k-ye) [reply taichi/ir/ir.h:88]: So, what kind of "global" variables is stamped by this identifier? I'm slightly against putting it into `Program`, as `Program` is more for runtime stuff. I.e., it's better to make AST module self contained.

Comment (PGZXB) [reply taichi/ir/ir.h:88]: "global" variables: global vars referenced by kernel, e.g. `a = ti.field(ti.i32, ())`, which don't belongs to a kernel but a `Program`, so maybe the 'id_counter' should be member of `Program` and rename it to `global_id_counter`. 


>it's better to make AST module self contained.

LGTM, but the global taichi objects are not built by `ASTBuilder`.

Comment (k-ye) [reply taichi/ir/ir.h:88]: I see. I think it's fine to move it to `Program` as a start, but long term speaking, we should have dedicated counters for each kind of objects. E.g. `SNode::id_counter`, `Field::id_counter`, etc.

Comment (PGZXB) [reply taichi/ir/ir.h:88]: Emmm... I have a immature idea. If my changes in the pr are merged, the `Identifier::id_counter` will only be used by `GlobalVariableExpression::GlobalVariableExpression(SNode *snode)` (as  `Identifier::id`). Can we use `snode->id` as `Identifier::id` and remove `Identifier::id_counter`?

Comment (k-ye) [reply taichi/ir/ir.h:88]: Hmm... I believe Ndarray also uses GlobalVariableExpression, and I'm not sure whether `GlobalVariableExpression` and `SNode` has a one-to-one mapping. Do  you remember? @strongoier @lin-hitonami

Comment (PGZXB) [reply taichi/ir/ir.h:88]: Very sorry, my bad, you are right. `GlobalVariableExpression` is used by `SNode` and others. 
I will move `id_counter` to Program and commit to the pr.

**Review by strongoier — APPROVED** [review]
Great! Thanks!

**Pull request merged**